numba/cpython/builtins.py

```diff
@@ -928,3 +929,77 @@ def impl(obj, name, default):
         return resolve_getattr(obj, name, default)
     return impl
 
+
+@intrinsic
+def resolve_hasattr(tyctx, obj, name):
+    if not isinstance(name, types.StringLiteral):
+        raise RequireLiteralValue("argument 'name' must be a literal string")
+    lname = name.literal_value
+    fn = tyctx.resolve_getattr(obj, lname)
+    # Whilst technically the return type could be a types.bool_, the literal
+    # value is resolvable at typing time. Propagating this literal information
+    # into the type system allows the compiler to prune branches based on a
+    # hasattr predicate. As a result the signature is based on literals. This is
+    # "safe" because the overload requires a literal string so each will be a
+    # different variant of (obj, literal(name)) -> literal(bool).
+    if fn is None:
+        retty = types.literal(False)
+    else:
+        retty = types.literal(True)
+    sig = retty(obj, name)
+    def impl(cgctx, builder, sig, ll_args):
+       return cgutils.false_bit if fn is None else cgutils.true_bit
+    return sig, impl
+
+# hasattr cannot be implemented as a getattr call and then catching
+# AttributeError because Numba doesn't support catching anything other than
+# "Exception", so lacks the specificity required. Instead this implementation
+# tries to resolve the attribute via typing information and returns True/False
+# based on that.
+@overload(hasattr)
+def ol_hasattr(obj, name):
+    def impl(obj, name):
+        return resolve_hasattr(obj, name)
+    return impl
```

Comment:
I think the implementation for `hasattr` is correct but it might produce false negatives in some cases, due to how Numba overloads some builtin functions. i.e. 

```python
@njit
def test():
    s = set([1, 2, 3])
    return hasattr(s, '__len__'), len(s)

print(test())
```